File: 01_Arrays_&_Hashing/06_Product_of_Array_Except_Self/0238-product-of-array-except-self.cpp
/*
Problem: LeetCode 238 - Product of Array Except Self

Description:
Given an integer array nums, return an array output such that output[i] is equal to the product of all the elements of nums except nums[i].

Intuition:
To find the product of all elements except self, we can utilize prefix and suffix products. By calculating the prefix product from the left and the suffix product from the right, we can obtain the desired result efficiently.

Approach:
1. Initialize an output array of the same size as the input array nums.
2. Calculate the prefix product from the left side of the array and store it in the output array.
3. Calculate the suffix product from the right side of the array and multiply it with the corresponding element in the output array.
4. Return the resulting output array.

Time Complexity:
The time complexity of this approach is O(n), where n is the size of the input array nums.
This is because we iterate through the array twice, once for calculating the prefix product and once for calculating the suffix product.

Space Complexity:
The space complexity is O(1), as we are reusing the input array for storing the output.
*/

class Solution {
  public:
    vector<int> productExceptSelf(vector<int> &nums) {
        int n = nums.size();
        vector<int> output(n, 1);
        // Calculate prefix product
        int prefix = 1;

        for (int i = 0; i < n; i++) {
            output[i] *= prefix;
            prefix *= nums[i];
        }

        // Calculate suffix product and multiply with prefix product stored in the output array
        int suffix = 1;

        for (int i = n - 1; i >= 0; i--) {
            output[i] *= suffix;
            suffix *= nums[i];
        }

        return output;
    }
};

// class Solution {
// public:
//     vector<int> productExceptSelf(vector<int>& nums) {
//         vector<int> res(nums.size(), 1);
//         // First, compute the prefix product and store in res
//         // res[i] = product of elements in nums from index 0, 1, ... to i - 1
//         for (int i = 0; i < nums.size() - 1; i++) {
//             res[i+1] = nums[i] * res[i];
//         }
//         // Second, compute the final result
//         int suffixProduct = 1;
//         for(int j = nums.size()-1; j > 0; j--) {
//             suffixProduct *= nums[j];
//             res[j-1] *= suffixProduct;
//         }
//         return res;
//     }
// };

File: 01_Arrays_&_Hashing/06_Product_of_Array_Except_Self/0238-product-of-array-except-self.rs
/*
Problem: LeetCode 238 - Product of Array Except Self

Key Idea:
The key idea is to compute the product of all elements to the left of each element and the product of all elements to the right of each element. The final product for each element is the product of these left and right products.

Approach:
1. Create two arrays, `prefix` and `postfix`, of the same size as the input array to store the products of elements to the left and right of each index, respectively.
2. Initialize `prefix[0] = 1` and `postfix[n-1] = 1`, where n is the length of the input array.
3. Populate the `prefix` array by iterating through the input array from left to right. `prefix[i]` will be the product of `prefix[i-1]` and `nums[i-1]`.
4. Populate the `postfix` array by iterating through the input array from right to left. `postfix[i]` will be the product of `postfix[i+1]` and `nums[i+1]`.
5. For each index i, the result for that index is `prefix[i] * postfix[i]`.

Time Complexity:
O(n), where n is the length of the input array. We iterate through the input array three times.

Space Complexity:
O(n), as we use two additional arrays, `prefix` and `postfix`, each of size n.
*/

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut prefix = vec![1; n];
        let mut postfix = vec![1; n];

        for i in 1..n {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        for i in (0..n - 1).rev() {
            postfix[i] = postfix[i + 1] * nums[i + 1];
        }

        let mut result = vec![1; n];
        for i in 0..n {
            result[i] = prefix[i] * postfix[i];
        }

        result
    }
}
